core/trino-main/src/main/java/io/trino/sql/planner/RelationPlanner.java

```diff
@@ -1865,9 +1865,31 @@ private SetOperationPlan process(SetOperation node)
         ImmutableListMultimap.Builder<Symbol, Symbol> symbolMapping = ImmutableListMultimap.builder();
         ImmutableList.Builder<PlanNode> sources = ImmutableList.builder();
 
-        for (Relation child : node.getRelations()) {
+        List<Relation> relations = node.getRelations();
+        checkArgument(relations.size() == 2, "relations size must be 2");
+        Relation rightRelation = relations.getLast();
+        for (Relation child : relations) {
             RelationPlan plan = process(child, null);
 
+            if (node.isCorresponding() && child.equals(rightRelation)) {
```

Comment:
The planner should not be doing any mappings based on names. That should have been established during analysis, and any necessary supporting data structure recorded in the Analysis object. E.g., a list of fields to select in the order in which they should be considered + the corresponding relation types.